### PR TITLE
QLoRA tutorial

### PR DESCRIPTION
**Pull request by rohan-varma**
#### Context
- Adds a QLoRA tutorial to accompany the LoRA tutorial in torchtune. Discussed with @ebsmothers, the nuances and details in QLoRA are enough for it to warrant its own tutorial. Will also link LoRA tutorial to QLoRA and backlink QLoRA to LoRA tutorial.

#### Changelog
- ...

#### Test plan
- 
image

### CODE REVIEW

File: docs/source/examples/qlora_finetune.rst
Comment (RdoubleA) [inline docs/source/examples/qlora_finetune.rst:32]: redundant on top of
```suggestion
memory efficiency. In LoRA, model parameters can be thought of as existing in two partitions: adapters, which are
```

File: docs/source/examples/qlora_finetune.rst
Comment (RdoubleA) [inline docs/source/examples/qlora_finetune.rst:37]: is this the correct usage of bespoke :)

Comment (RdoubleA) [inline docs/source/examples/qlora_finetune.rst:37]:
```suggestion
QLoRA further quantizes the base model parameters into a bespoke 4-bit NormalFloat (NF4) data type, resulting in 4-8x less parameter memory usage while
```

File: docs/source/examples/qlora_finetune.rst
Comment (RdoubleA) [inline docs/source/examples/qlora_finetune.rst:44]: 🤯

File: docs/source/examples/qlora_finetune.rst
Comment (RdoubleA) [inline docs/source/examples/qlora_finetune.rst:48]: I have mixed feelings about this section, I think framing it as diving into core abstractions is a bit too technical for a tutorial and a lot of the code might be too much information. Most users just want a high level understanding of what QLoRA is and how they can use it in torchtune. I think the example code showing the cuda memory allocated differences is awesome, and maybe you can keep a demo of the linear_nf4 but in a more contrived example that shows that the quantized weight is saved for backward.

Or maybe another way to organize it: high level of what is QLoRA, how do you run it in torchtune, show the memory differences, and use the memory allocated / linear_nf4 examples to demonstrate how memory is saved. This way, you introduce the concepts from high level and easy to understand to more specific and more nitty gritty.

Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:48]: So I mostly agree, though I do think tutorials, in addition to clearly showcasing how the feature can be used in TorchTune, can also add value by deep diving into how things work under the hood to help with user understanding. If users purely want to use QLoRa for a finetune, they can do so at a high level via the README. 

For this particular section, I tried to address this concern by having the first portion of this section be pretty high level - how to just pass in `quantize_base` flag to LoRALinear and get a QLoRA enabled linear. The rest of the section goes deeper into how it actually happens - maybe can add a "this part is optional" flag or something like that.

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:49]: Personally I think this heading can be a bit more concrete. Why not "How QLoRA saves memory in nn.Linear" or something?

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:163]:
```suggestion
Under the hood, this will apply LoRA to the ``q_proj`` and ``v_proj`` matrices in all attention layers, and further quantize the base parameters
```

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:56]: (Putting this comment here kinda arbitrarily) One thing I don't see mentioned anywhere is "compute datatype" and "storage datatype". I've definitely seen this description elsewhere (maybe in the original paper?) and personally I feel like it's a bit more clear than writing NF4, bf16/fp32 all the time (though obv you need to mention the exact datatypes in some places)

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:177]: nit: too much interoperate

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:185]: Personally I think this whole section could be more easily illuminated via a code block with e.g. .state_dict() with peak memory printed without the state dict hook, then add the hook and do the same thing.

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:252]: I'm a bit torn here.. while I think it's nice to print the raw output that someone will see in the logs, personally I would just do it one time, e.g. "you will see something like this" and then put it in a table (QLoRA, LoRA) x (peak init memory, peak training memory). Then it's a bit more digestible

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:193]: This hyperlink format is so annoying lol
```suggestion
with a `QLoRA configuration <https://github.com/pytorch/torchtune/blob/main/recipes/configs/llama2/7B_qlora_single_device.yaml>`_.
```

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:200]: Should be able to scrap the `recipes/configs/` part here, right? (Same comment for the LoRA command below)

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:264]: nit: don't say users

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:278]: Btw you can also add explicit labels to the two lines in the figure, e.g. iconic-pasma-57 -> LoRA and azure-bird-56 -> QLoRA

File: docs/source/examples/qlora_finetune.rst
Comment (RdoubleA) [reply docs/source/examples/qlora_finetune.rst:43]: nah, I think this is fine. you don't know which paragraph someone will start reading at if they're skimming or jumping around

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [reply docs/source/examples/qlora_finetune.rst:48]: I think the optional part makes sense here. 

The other thing I would recommend just in terms of format is just to switch it up between code blocks and text (and if possible images) a bit, especially on the denser topics. Maybe I have a short attention span, but once I read >=3 sentences of upcast/downcast/dtype talk my eyes start to glaze over, and changing the medium you're conveying information with is a good way to keep things engaging. 

Also can be worth thinking about if there are places where it's OK to be less than 100% precise for the sake of getting the essential idea across, then point to a reference elsewhere. (Admittedly I already looked through here for places to do that, and I think the subject matter just makes it a bit hard tbh.. it's often either all the gory details or none at all)

Comment (RdoubleA) [reply docs/source/examples/qlora_finetune.rst:48]: Don't like flagging a whole section as optional, that feels odd to me. How about just changing the order of these sections and making this section a little more succinct? Because the meat is how do we run QLoRA in torchtune, and that should be higher up.

File: docs/source/examples/qlora_finetune.rst
Comment (joecummings) [reply docs/source/examples/qlora_finetune.rst:43]: I think you should link every other reference to the paper to keep people on their toes.

File: docs/source/examples/qlora_finetune.rst
Comment (joecummings) [reply docs/source/examples/qlora_finetune.rst:48]: Can we move this to a Deep Dive article?

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:37]: I usually take bespoke in an eng context to me custom made, so IMO yes, because it was specifically built for QLoRA (not by us, but by paper authors)

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:41]: Definitely interested! Will punt this out to a follow up though.

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:44]: IMO its too in the weeds and not worth it to just re explain if its in the paper. Can add a line directing folks to the paper, but IMO its already clear enough to read the paper for this sort of detail

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:45]: Done.

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:48]: Ok, I moved this to the bottom right now to focus on using QLoRA in torchtune and leaving this as a deep dive if folks are interested. Can move it to a separate article in a follow up PR if that's an even better approach.

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:49]: Made it Using QLoRA to save memory

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:56]: Added a blurb about compute vs storage dtype right before this.

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:80]: done! thanks for the tip

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:119]: changed this -> frozen_out

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:142]: Could potentially add this in follow up work, but tbh I would defer to torchao to explain this more rigorously. Can add an explanation in the deep dive section that's now at the end if that's helpful.

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:200]: Done

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:32]: nit: multiple instances of "on top of LoRA" is repetitive

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [reply docs/source/examples/qlora_finetune.rst:37]: Personally I also don't like it, but for a different reason. To me it implies something that is custom or one-off, i.e. not general. In that case I think it conveys the wrong message about NF4

File: docs/source/examples/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/examples/qlora_finetune.rst:252]: Added table

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [reply docs/source/examples/qlora_finetune.rst:45]: Sorry one more nit here, but I think we are going with torchao (all lowercase) now

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:9]: nit: you should find and replace TorchTune -> torchtune as we have done it everywhere else since after you first opened this PR

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:34]: nit
```suggestion
the original model. In vanilla LoRA-style training, both these parameters are held in the same precision (typically fp32 or bf16), and
```

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:46]: nit: I think we want TorchAO -> torchao

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:113]: nit. Also might remove checkpoint conversion since idk that it's really an ecosystem thing in the same way the other items are
```suggestion
TorchTune and beyond (e.g. checkpoint conversion, post-training quantization, evaluation, inference). This conversion process also allows LoRA adapter weights to be merged back into the base model as done
```

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:119]: nit: this sentence could be a little unclear, seemingly implying that the way we avoid peaking memory is by maintaining an entire bf16/fp32 copy on GPU.

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst:183]:
```suggestion
    tune run lora_finetune_single_device --config llama2/7B_qlora_single_device.yaml compile=True
```

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [reply docs/source/examples/qlora_finetune.rst]: Sorry to be annoying, but can you filter the x-axis to [0, 1000] or something in wandb and reupload? Otherwise it looks weird that one is running longer.

File: docs/source/examples/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/examples/qlora_finetune.rst]: You can remove this now that you've added the legend
```suggestion
A comparison of the smoothed loss curves between QLoRA and LoRA can be seen below.
```

File: docs/source/tutorials/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/tutorials/qlora_finetune.rst:203]: Add a section header here? (Maybe something like "Appendix: Building QLoRA from LoRA")

File: docs/source/tutorials/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/tutorials/qlora_finetune.rst:210]: 😍

File: docs/source/tutorials/qlora_finetune.rst
Comment (ebsmothers) [inline docs/source/tutorials/qlora_finetune.rst:275]: No longer applicable

File: docs/source/tutorials/qlora_finetune.rst
Comment (ebsmothers) [reply docs/source/tutorials/qlora_finetune.rst:203]: Would also maybe add some transition into the appendix at the end of the previous section.

**Review by ebsmothers — APPROVED** [review]
OK left a handful of other comments, please make sure they're addressed before landing. Modulo that I think this is looking good!

File: docs/source/tutorials/qlora_finetune.rst
Comment (rohan-varma) [reply docs/source/tutorials/qlora_finetune.rst:205]: DOne

**Pull request merged**
